crates/sui-json-rpc/src/error.rs

```diff
@@ -84,34 +85,133 @@ impl Error {
-    ContainsDuplicates,
+pub enum ServerError {
+    #[error("Serde error")]
+    Serde(SuiError),
 
-    #[error("Input exceeds limit of {0}")]
-    SizeLimitExceeded(String),
+    #[error("Bcs error")]
+    Bcs(bcs::Error),
+}
 
-    #[error("{0}")]
-    GenericNotFound(String),
+impl From<ServerError> for RpcError {
+    fn from(_e: ServerError) -> Self {
+        to_internal_error("Internal server error, please try again later")
+    }
+}
 
-    #[error("{0}")]
-    GenericInvalid(String),
+fn to_internal_error(err: impl ToString) -> RpcError {
+    let error_object = ErrorObject::owned(-32603, err.to_string(), None::<String>);
+    RpcError::Call(CallError::Custom(error_object))
+}
+
+// One-to-one of the return type of the RPC method where applicable
+#[derive(Debug, Clone, Copy, AsRefStr, EnumString)]
+pub enum EntityType {
+    DelegatedStake,
+    SuiMoveNormalizedStruct,
+    SuiMoveNormalizedFunction,
+    Object,
+    Package,
+    Product,
+    SuiMoveNormalizedModule,
+    Checkpoint,
+}
+
+impl fmt::Display for EntityType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self.as_ref())
+    }
+}
+
+#[derive(Debug, Error)]
+pub enum ClientError {
+    // Any kind of serialization or deserialization error
+    #[error("Serialization error on '{param}': {reason}")]
+    Serde { param: &'static str, reason: String },
+
+    #[error("Base64 error on '{param}': {reason}")]
+    Base64 { param: &'static str, reason: String },
+
+    #[error("BCS error on '{param}': {reason}")]
+    Bcs {
+        param: &'static str,
+        reason: bcs::Error,
+    },
+
+    #[error("FastCrypto error on '{param}': {reason}")]
+    FastCrypto {
+        param: &'static str,
+        reason: FastCryptoError,
+    },
 
-    #[error("request_type` must set to `None` or `WaitForLocalExecution` if effects is required in the response")]
+    #[error("Invalid combination of type and value: {0}")]
+    SerdeWithLayout(String),
+
+    #[error(transparent)]
+    Domain(#[from] DomainParseError),
+
+    #[error("Invalid '{param}': {reason}")]
+    InvalidParam { param: &'static str, reason: String },
+    // maybe InvalidParamMulti or something. param, value, reason
+    #[error("`request_type` must set to `None` or `WaitForLocalExecution` if effects is required in the response")]
     InvalidExecuteTransactionRequestType,
 
-    #[error("Unsupported protocol version requested. Min supported: {0}, max supported: {1}")]
-    ProtocolVersionUnsupported(u64, u64),
+    #[error("{entity} '{id}' not found")]
+    NotFound { entity: EntityType, id: String },
+
+    #[error("{entity} '{id}' not found in {in_entity} '{in_id}'")]
+    NotFoundIn {
+        entity: EntityType,
+        id: String,
+        in_entity: EntityType,
+        in_id: String,
+    },
 
-    #[error("Unable to serialize: {0}")]
-    CannotSerialize(#[from] bcs::Error),
+    #[error("{entity} '{id}' not found most likely due to pruning")]
```

Comment:
Could not find 
```suggestion
    #[error("Could not find {entity} '{id}'. This is most likely due to pruning.")]
```

What can one do to overcome this issue?